lib/API/DX/Device.cpp

```diff
@@ -521,51 +528,115 @@ class DXDevice : public offloadtest::Device {
                   IID_PPV_ARGS(&UploadBuffer)),
               "Failed to create committed resource (upload buffer)."))
         return Err;
 
-      // Initialize the SRV data
+      // Tile mapping setup (only skipped when TilesMapped is set to 0)
+      const UINT NumTiles = getNumTiles(R.TilesMapped, ResDesc.Width);
+      ComPtr<ID3D12Heap> Heap; // optional, only created if NumTiles > 0
+
+      if (NumTiles > 0) {
+        std::vector<D3D12_TILED_RESOURCE_COORDINATE> StartCoords(NumTiles);
+        std::vector<D3D12_TILE_REGION_SIZE> RegionSizes(NumTiles);
+        std::vector<D3D12_TILE_RANGE_FLAGS> RangeFlags(
+            NumTiles, D3D12_TILE_RANGE_FLAG_NONE);
+        std::vector<UINT> HeapRangeStartOffsets(NumTiles);
+        std::vector<UINT> RangeTileCounts(NumTiles, 1);
+
+        // Create a Heap large enough for the mapped tiles
+        D3D12_HEAP_DESC HeapDesc = {};
+        HeapDesc.Properties = CD3DX12_HEAP_PROPERTIES(D3D12_HEAP_TYPE_DEFAULT);
+        HeapDesc.Alignment = D3D12_DEFAULT_RESOURCE_PLACEMENT_ALIGNMENT;
+        HeapDesc.SizeInBytes = static_cast<UINT64>(NumTiles) *
+                               D3D12_DEFAULT_RESOURCE_PLACEMENT_ALIGNMENT;
+        HeapDesc.Flags = D3D12_HEAP_FLAG_ALLOW_ALL_BUFFERS_AND_TEXTURES;
+
+        if (auto Err =
+                HR::toError(Device->CreateHeap(&HeapDesc, IID_PPV_ARGS(&Heap)),
+                            "Failed to create heap for tiled SRV resource."))
+          return Err;
+
+        // Fill tile coordinates and region sizes
+        for (UINT I = 0; I < NumTiles; ++I) {
+          StartCoords[I] = {I, 0, 0, 0};
+          RegionSizes[I].NumTiles = 1;
+          RegionSizes[I].UseBox = FALSE;
+          HeapRangeStartOffsets[I] = I;
+        }
+
+        // Retrieve a command queue from InvocationState
+        ID3D12CommandQueue *CommandQueue = IS.Queue.Get();
+
+        // Map the first NumTiles tiles in the Buffer
+        CommandQueue->UpdateTileMappings(
+            Buffer.Get(), NumTiles, StartCoords.data(), RegionSizes.data(),
+            Heap.Get(), NumTiles, RangeFlags.data(),
+            HeapRangeStartOffsets.data(), RangeTileCounts.data(),
+            D3D12_TILE_MAPPING_FLAG_NONE);
+      }
+
+      // Upload data initialization
       void *ResDataPtr = nullptr;
-      if (auto Err = HR::toError(UploadBuffer->Map(0, nullptr, &ResDataPtr),
-                                 "Failed to acquire UAV data pointer."))
-        return Err;
-      memcpy(ResDataPtr, ResData.get(), R.size());
-      UploadBuffer->Unmap(0, nullptr);
+      const D3D12_RANGE Range = {0, 0}; // no reads expected
+      if (SUCCEEDED(UploadBuffer->Map(0, &Range, &ResDataPtr))) {
+        memcpy(ResDataPtr, ResData.get(), R.size());
+        UploadBuffer->Unmap(0, nullptr);
+      } else {
+        return llvm::createStringError(std::errc::io_error,
+                                       "Failed to map SRV upload buffer.");
+      }
 
+      // Add GPU upload commands
```

Comment:
This comment adds no value since the name of the function is pretty clear about what it does.